pkg/ingester/wal.go

```diff
@@ -0,0 +1,726 @@
+package ingester
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log/level"
+	"github.com/golang/protobuf/proto"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+	"github.com/prometheus/prometheus/tsdb/wal"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+// WALConfig is config for the Write Ahead Log.
+type WALConfig struct {
+	walEnabled         bool
+	checkpointEnabled  bool
+	recover            bool
+	dir                string
+	checkpointDuration time.Duration
+	metricsRegisterer  prometheus.Registerer
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *WALConfig) RegisterFlags(f *flag.FlagSet) {
+	f.BoolVar(&cfg.walEnabled, "ingester.wal-enable", false, "Enable the WAL.")
+	f.BoolVar(&cfg.checkpointEnabled, "ingester.checkpoint-enable", false, "Enable checkpointing.")
+	f.BoolVar(&cfg.recover, "ingester.recover-from-wal", false, "Recover data from existing WAL.")
+	f.StringVar(&cfg.dir, "ingester.wal-dir", "", "Directory to store the WAL.")
+	f.DurationVar(&cfg.checkpointDuration, "ingester.checkpoint-duration", 1*time.Hour, "Duration over which to checkpoint.")
+}
+
+// WAL interface allows us to have a no-op WAL when the WAL is disabled.
+type WAL interface {
+	Log(record *Record) error
+	Stop()
+}
+
+type noop struct{}
+
+// Log a Record to the WAL.
+func (noop) Log(*Record) error {
+	return nil
+}
+
+// Stop any background WAL processes.
+func (noop) Stop() {}
+
+type walWrapper struct {
+	cfg  WALConfig
+	quit chan struct{}
+	wait sync.WaitGroup
+
+	wal           *wal.WAL
+	getUserStates func() map[string]*userState
+
+	// Checkpoint metrics.
+	checkpointDeleteFail    prometheus.Counter
+	checkpointDeleteTotal   prometheus.Counter
+	checkpointCreationFail  prometheus.Counter
+	checkpointCreationTotal prometheus.Counter
+}
+
+// newWAL creates a WAL object.
+// * If the WAL is disabled, then the returned WAL is a no-op WAL.
+// * If WAL recovery is enabled, then the userStates is always set for ingester.
+func newWAL(cfg WALConfig, userStatesFunc func() map[string]*userState) (WAL, error) {
+	if !cfg.walEnabled {
+		return &noop{}, nil
+	}
+
+	var walRegistry prometheus.Registerer
+	if cfg.metricsRegisterer != nil {
+		walRegistry = prometheus.WrapRegistererWith(prometheus.Labels{"kind": "wal"}, cfg.metricsRegisterer)
+	}
+	tsdbWAL, err := wal.NewSize(util.Logger, walRegistry, cfg.dir, wal.DefaultSegmentSize/4, true)
```

Comment:
I have reduced the default segment size as it allows for fine-grained deletion of segment files after checkpointing and possible reduction of replay time.